### ISSUE TITLE
Add lighting to the engine.

### ISSUE THREAD

**Issue by DhirajWishal** [DhirajWishal/Xenon#8]
As of now, the engine doesn't have any concept of lighting. This is not good. For this, my first plan is to add a structure or class called light which has the following properties.

* Intensity
* Color
* Position
* Direction
* Field angle

I'm sure this information can be stored in a single structure like so.
```c++
struct Light final 
{
    glm::vec4 m_Color;
    glm::vec3 m_Position;
    glm::vec3 m_Direction;
    float m_Intensity = 1.0f;    // 0 - No intensity, 1 - Full intensity.
    float m_FieldAngle = 15;
};
```

The lighting can be added to the scene descriptor at a specific binding. We also need to add support for having more than one light source.

**Comment by DhirajWishal** [comment]
I decided to extend the engine a bit more to add support for this and to make future implementations much easier. I added a new class called `Scene` to the engine which stores information like models, lighting, camera and a bunch more. Essentially the models and everything that passes using the scene descriptor.

I also added ECS to the engine using [entt](https://github.com/skypjack/entt). I decided not to use the good old entity-component mapping to rather use the term grouping. They're just entities in Xenon. The logic behind it is that we can group components to form a grouping. And a grouping can have other things which we might add in the future like audio, triggers and stuff. Grouping can have sub-grouping which allows us to have a hierarchy.

**Comment by DhirajWishal** [comment]
We might have to detach the descriptors from the pipelines and allow the user or the engine to create them using bindings. This also means descriptors have to be managed externally (which might be a pain). I think it would be best to expose the descriptor [set] manager to the frontend and create descriptors from there.

**Comment by DhirajWishal** [comment]
Okay! So lighting was added but not the way you expected it to be. 

Lighting exists as a struct (component) that stores information about light. That's all that we do now. We had to improve the engine quite a bit to make it more flexible. But that's all we did because I am looking into extending the engine a little more than just shadow maps. I want to look into GI. I'm not sure about what the algorithm will be but either way, it's gonna be a huge task and the engine will be changed for the best.

So there you go. We have lighting!
